### ISSUE TITLE
Hebrew texts renders as Arabic!

### ISSUE THREAD

**Issue by danielrosehill** [autokey/autokey#207]
## Classification:

Bug

## Reproducibility:

Always

## Summary

When creating a phrase macro in Hebrew, and using the keyboard paste method, the text renders in nonsensical Arabic characters.

## Steps to Reproduce (if applicable)

1. Create a phrase macro in Hebrew (type or copy and paste from here: https://he.wikipedia.org/wiki/%D7%A2%D7%9E%D7%95%D7%93_%D7%A8%D7%90%D7%A9%D7%99)

2. Assign a hotkey and use the keyboard method.

## Expected Results

Macro pastes

## Actual Results

Using the hotkey will result in the insertion of random Arabic text.

Creating macro:
![selection_001](https://user-images.githubusercontent.com/40233203/47878920-10665480-de28-11e8-8036-a9d67224f911.png)

Result:
![selection_002](https://user-images.githubusercontent.com/40233203/47878931-14927200-de28-11e8-8046-83393ed80abc.png)
```

## Version

AutoKey version: 0.90.4

Used GUI (Gtk, Qt, or both): GUI

If the problem is known to be present in more than one version, please list all of those.

Installed via: command line from respository. 

Distro: Lubuntu 18.04 LTS.

**Comment by josephj11** [comment]
There are several issues here, but they are probably all version-related.

First, you are using 0.90.4 which is almost seven years old and unsupported. This version had a lot of problems with Unicode character sets and was based on Python 2.7. Unfortunately, a few distros (I'm talking to you, Ubuntu) still package it.
The newer versions of AutoKey are based on Python 3.x which has much better support for multibyte character sets and there have been a large number of bug fixes and a few major enhancements.
So, the first answer is, don't use 0.90.4!

Here are [instructions](https://github.com/autokey/autokey/wiki/Installing#debian-and-derivatives) on how to install a newer version of AutoKey.

Before you do that:
Everything below is done on a terminal such as konsole or xterm unless
explicitly noted. If you're not familiar with the Linux CLI (command
line interface and bash), we can take each step in full detail so you'll
get it right.

You need to backup everything under ~/.config/autokey - preferably by
just renaming it something like ~/.config/autokey_old

and then

sudo apt purge autokey

That will get you a clean slate, but any macros and phrases you have
defined will still be in your backup (and we can restore them later once
the new version is installed and working.)

**Comment by danielrosehill** [comment]
Hi again Joe!

I upgraded to the newer version (0.951.1) using your instructions, but both Unicode character issues (the inability of the em dash macro to work and Hebrew text rendering as nonsenical Arabic) persist.

**Comment by danielrosehill** [comment]
.... and yes, as I mentioned in the log, I am using Lubuntu. At least I have a newer version now (and the crashing issue seems to have been entirely resolved, which is great)!

**Comment by josephj11** [comment]
Now you have the general release version. Since the problem is still there, you'll need to get the most recent version [0.95.4](https://github.com/autokey/autokey/releases/tag/v0.95.4). You will have to install it manually, so it won't automatically get updates. If your bug persists in that, then it's still a current issue.
Download the common and gtk debs to a directory on your system
Change into that directory (in a terminal) then
To install it, use the following:
`sudo dpkg --install autokey-common_0.95.4-0_all.deb autokey-gtk_0.95.4-0_all.deb`

They need to be installed at the same time and `dpkg` will get it right while `apt` might not.
If you're using the QT front end, you can download and install that instead of the GTK version.
If it still doesn't work, then a developer will have to look at it.

**Comment by luziferius** [comment]
@josephj11 Well, that wont be needed. The newest version did not change anything that could magically fix this issue.

@drjlm You may not be able to _type_ hebrew text using autokey at all, until the X11 interface is re-written. Non-latin languages cause trouble when used with autokey.
Do you use multiple keyboard layouts, or have multiple ones configured? This may cause autokey to type using the wrong keyboard configuration/layout. (Personally, I run into this when `neo2` is configured as the default layout and I switch to something else.)

If not, you may want to try the clipboard pasting method. This method uses the clipboard and bypasses the need to type the letters. So in general, the clipboard method should work regardless of the phrase content.

**Comment by danielrosehill** [comment]
@luziferius Yes, I have English and Hebrew as system layouts - but not Arabic. I tried all the possible methods (including the two clipboard pasting ones) and both yielded the same result (Hebrew text rendering as Arabic; em dash not entering). Don't think this is important enough to spend more time on, but hope that the Unicode issue is resolved in future updates.

**Comment by luziferius** [comment]
That sounds like keyboard input is mangled (maybe twice) by the system.
Autokey types using the base keyboard layout, because that is what your physical keyboard is configured to, regardless of the choice made in the tray icon selector. (On KDE, it is the first in the configuration window list). These keystrokes are then translated to the active layout (somewhat like the scan-code to key-code translation). The translated output is then again translated using the same translation table. In the first step, it translates 8-bit ASCII to some multi-byte UTF8 (or so) that is needed for Hebrew. Then each is translated again (maybe even byte-wise?). I guess the result happens to be valid Arabic UTF-8 characters…

I get something similar on my machine. Try to remove the English layout for testing purposes. (And to be safe, restart autokey after you removed the layout).
Then try to type Hebrew again. does it still happen?
